Canvas' security should not take into account Single Origin (for videos and images)

jyavenard opened the pull request:
#### 2c5193d0471c2417232a02a7aff688e4744f5839
<pre>
Canvas&apos; security should not take into account Single Origin (for videos and images)
<a href="https://bugs.webkit.org/show_bug.cgi?id=242889">https://bugs.webkit.org/show_bug.cgi?id=242889</a>
rdar://97245327

Reviewed by Youenn Fablet.

Per canvas&apos; spec:
<a href="https://html.spec.whatwg.org/multipage/canvas.html#the-image-argument-is-not-origin-clean">https://html.spec.whatwg.org/multipage/canvas.html#the-image-argument-is-not-origin-clean</a>

an object image is not origin-clean if:
HTMLOrSVGImageElement
    image&apos;s current request&apos;s image data is CORS-cross-origin.
HTMLVideoElement
    image&apos;s media data is CORS-cross-origin.
HTMLCanvasElement
ImageBitmap
    image&apos;s bitmap&apos;s origin-clean flag is false.

And as per the security&apos;s policy:
<a href="https://html.spec.whatwg.org/multipage/canvas.html#security-with-canvas-elements">https://html.spec.whatwg.org/multipage/canvas.html#security-with-canvas-elements</a>

&quot;To mitigate this, bitmaps used with canvas elements and ImageBitmap objects are defined to have a flag indicating whether they are origin-clean. All bitmaps start with their origin-clean set to true. The flag is set to false when cross-origin images are used.&quot;

The implementation prevented drawing into a canvas a video that was served across multiple mirrors as is commonly found in the media world.

* Source/WebCore/html/canvas/CanvasRenderingContext.cpp:
(WebCore::CanvasRenderingContext::wouldTaintOrigin):
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::isCORSCrossOrigin const):
* Source/WebCore/loader/cache/CachedResource.h:

Canonical link: <a href="https://commits.webkit.org/257207@main">https://commits.webkit.org/257207@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35fa92118da68ea41a719d450f4203a123f1a461

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98111 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7325 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31260 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107572 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167836 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102051 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7814 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36089 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90707 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104177 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5873 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84707 "Built successfully") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32998 "Found 2 new test failures: http/tests/security/canvas-remote-read-remote-video-redirect.html, imported/w3c/web-platform-tests/html/semantics/embedded-content/the-canvas-element/security.pattern.fillStyle.sub.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87742 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89457 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1296 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1257 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22392 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4964 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6122 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2563 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41804 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -627,7 +627,7 @@ static void logResourceLoaded(Frame* frame, CachedResource::Type type)
 Expected<void, String> SubresourceLoader::checkResponseCrossOriginAccessControl(const ResourceResponse& response)
 {
-    if (!m_resource->isCrossOrigin() || options().mode != FetchOptions::Mode::Cors)
+    if (!m_resource->isCORSCrossOrigin())
```
I do not think this is correct here.
We want to have mode CORS and the load to be cross origin.
CachedResource::isCORSCrossOrigin() is basically mode=no-cors and load is cross origin.

jyavenard replied:
indeed. I thought the test ended up being the same; but I obviously got that wrong

youennf commented on Source/WebCore/loader/cache/CachedResource.cpp:
```diff
@@ -394,6 +394,11 @@ bool CachedResource::isCrossOrigin() const
 
+bool CachedResource::isCORSCrossOrigin() const
+{
+    return m_responseTainting != ResourceResponse::Tainting::Basic && m_responseTainting != ResourceResponse::Tainting::Cors;
```
It might be better to align with the spec strictly by checking opaque or opaqueredirect.

jyavenard replied:
ok

youennf commented on Source/WebCore/loader/cache/CachedResource.cpp:
```diff
@@ -394,6 +394,11 @@ bool CachedResource::isCrossOrigin() const
     return m_responseTainting != ResourceResponse::Tainting::Basic;
 }
 
+bool CachedResource::isCORSCrossOrigin() const
```
It is not used anywhere, let's only add it if we plan to use it.

jyavenard replied:
it was supposed to be used, it's an error that it's not.

youennf commented on Source/WebCore/html/canvas/CanvasRenderingContext.cpp:
```diff
@@ -151,10 +153,7 @@ bool CanvasRenderingContext::wouldTaintOrigin(const HTMLVideoElement* video)
-        return true;
-
-    if (!(video->player() && video->player()->didPassCORSAccessCheck()) && video->wouldTaintOrigin(*m_canvas.securityOrigin()))
+    if (!video->didPassCORSAccessCheck() && video->wouldTaintOrigin(*m_canvas.securityOrigin()))
```
Can you remove the comment at the top of CanvasRenderingContext::wouldTaintOrigin?

Looking at wouldTaintOrigin implementation, I wonder whether the following case is actually handled correctly:
- Make a no cors video load to a cross origin
- The cross origin response is a redirect to a same origin resource
I would expect the video to be considered as cross origin but wouldTaintOrigin might return false there since it does not seem to be concerned about the redirect response origin.
This might be worth a test case if we do not already have one.

jyavenard replied:
The handling for videos is incorrect, it isn't based on CORS cross-origin data, but purely on the origin itself.
I opened https://bugs.webkit.org/show_bug.cgi?id=242889 to properly fix it in a follow-up

youennf commented on Source/WebCore/html/canvas/CanvasRenderingContext.cpp:
```diff
@@ -151,10 +153,7 @@ bool CanvasRenderingContext::wouldTaintOrigin(const HTMLVideoElement* video)
     if (!video || !m_canvas.originClean())
```
This above comment seems outdated.

youennf commented on Source/WebCore/html/canvas/CanvasRenderingContext.cpp:
```diff
@@ -151,10 +153,7 @@ bool CanvasRenderingContext::wouldTaintOrigin(const HTMLVideoElement* video)
     if (!video || !m_canvas.originClean())
         return false;
 
-    if (!video->hasSingleSecurityOrigin())
```
Can we remove HTMLVideoElement:: hasSingleSecurityOrigin() and related, or do you plan to do it in a follow-up?

jyavenard replied:
I will take it in a follow-up

jyavenard committed:
Canvas' security should not take into account Single Origin (for videos and images)
https://bugs.webkit.org/show_bug.cgi?id=242889
rdar://97245327

Reviewed by Youenn Fablet.

Per canvas' spec:
https://html.spec.whatwg.org/multipage/canvas.html#the-image-argument-is-not-origin-clean

an object image is not origin-clean if:
HTMLOrSVGImageElement
    image's current request's image data is CORS-cross-origin.
HTMLVideoElement
    image's media data is CORS-cross-origin.
HTMLCanvasElement
ImageBitmap
    image's bitmap's origin-clean flag is false.

And as per the security's policy:
https://html.spec.whatwg.org/multipage/canvas.html#security-with-canvas-elements

"To mitigate this, bitmaps used with canvas elements and ImageBitmap objects are defined to have a flag indicating whether they are origin-clean. All bitmaps start with their origin-clean set to true. The flag is set to false when cross-origin images are used."

The implementation prevented drawing into a canvas a video that was served across multiple mirrors as is commonly found in the media world.

* Source/WebCore/html/canvas/CanvasRenderingContext.cpp:
(WebCore::CanvasRenderingContext::wouldTaintOrigin):
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::isCORSCrossOrigin const):
* Source/WebCore/loader/cache/CachedResource.h:

Canonical link: https://commits.webkit.org/257207@main
